samza-api/src/main/java/org/apache/samza/metadatastore/MetadataStore.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.metadatastore;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.config.Config;
+import org.apache.samza.metrics.MetricsRegistry;
+import java.util.Map;
+
+/**
+ * Store abstraction responsible for managing the metadata of a Samza job.
+ */
+@InterfaceStability.Evolving
+public interface MetadataStore {
```

Comment:
Should there also be a `create()` API? Without it, this assumes the underlying store is already created. In the case of the case of the `CoordinatorStreamStore` implementation, the underlying stream gets created directly from the SystemAdmin here:

https://github.com/apache/samza/blob/06488bf7a7dfdfa6d26ed91f79b6c316460960b1/samza-core/src/main/scala/org/apache/samza/job/JobRunner.scala#L93

Should we replace that with a `MetadataStore#create()` call to fully abstract away the coordinator stream?

Reply:
1. I was thinking of introducing a new abstraction `MetadataStoreAdmin` with methods like `delete(store)`, `create(store)`, `validate`, `clear(store)` that will house all the admin related operations for a kind of metadata store. 
2. We will have two implementations viz `CoordinatorStreamMetadatastoreadmin` , `ZKMetadataStoreAdmin`  for yarn and standalone respectively. The existing `MetadataStore` contact assumes the presence of store and provides operations on it. IMHO, the admin operations should be kept separate abstraction. We should have `MetadataStoreFactory.getMetadataStoreAdmin(config, registry)` method and that should return admin object for a type of metadata store.

@prateekm @dnishimura  What do you guys think?

Reply:
1. I was thinking of introducing a new abstraction `MetadataStoreAdmin` with methods like `delete(store)`, `create(store)`, `validate`, `clear(store)` that will house all the admin related operations for a kind of metadata store.
2. We will have two implementations viz `CoordinatorStreamMetadatastoreadmin`, `ZKMetadataStoreAdmin` for yarn and standalone respectively. The existing `MetadataStore` contact assumes the presence of store and provides operations on it. IMHO, the admin operations should be kept separate and should be part of separate abstraction. We should have `getMetadataStoreAdmin(config, registry)` method in `MetadataStoreFactory` and that should return admin object for a type of metadata store.

@prateekm @dnishimura What do you guys think?

Reply:
It aligns with the `SystemFactory` pattern. Works for me.